### ISSUE TITLE
Trial tracebacks are not properly trimmed since 12.3.0

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6282]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)| @ralphm reported|
|-|-|
|Trac ID|trac#6282|
|Type|release blocker: regression|
|Created|2013-01-28 13:00:33Z|
|Branch|https://github.com/twisted/twisted/tree/async-trim-frames-6282|

Due to the porting effort of trial to Python 3, tracebacks in the trial reporter are no longer properly trimming tracebacks:

```python
from twisted.trial import unittest

class Test(unittest.TestCase):

    def test_it(self):
        raise Exception()
```

yields:

```
[..]
[ERROR]
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/twisted/internet/defer.py", line 138, in maybeDeferred
    result = f(*args, **kw)
  File "/usr/lib/python2.7/dist-packages/twisted/internet/_utilspy3.py", line 41, in runWithWarningsSuppressed
    reraise(exc_info[1], exc_info[2])
  File "/usr/lib/python2.7/dist-packages/twisted/internet/_utilspy3.py", line 37, in runWithWarningsSuppressed
    result = f(*a, **kw)
  File "/home/ralphm/work/wokkel/trunk/q.py", line 6, in test_it
    raise Exception()
exceptions.Exception:
[..]
```

This is due to importing `runWithWarningsSuppressed` from `_utilspy3` instead of `util` and an additional frame there. This makes the frames not being trimmed by `Reporter._trimFrames`.

Related tickets: [#6033](https://github.com/twisted/twisted/issues/6033) and [#6235](https://github.com/twisted/twisted/issues/6235).

<details><summary>Searchable metadata</summary>

```
trac-id__6282 6282
type__release_blocker__regression release blocker: regression
reporter__ralphm ralphm
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_async_trim_frames_6282 branches/async-trim-frames-6282
branch_author__therve therve
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1359378033000000 1359378033000000
changetime__1361435643000000 1361435643000000
version__None None
owner__therve therve
cc__jml
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

(In [36988]) Branching to 'async-trim-frames-6282'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve removed owner|
|-|-|

Nice catch, fix ready to review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

Can you comment on why the branch changes the assertions made in `test_hiddenException`?  It seems like that's a general improvement (?) in behavior, rather than a fix for a regression.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve commented|
|-|-|

Sure: in r35861 which introduced the regression by creating utilspy3, the test was changed in a wrong way to comply to the new behavior, instead of fixing it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

> Sure: in r35861 which introduced the regression by creating utilspy3, the test was changed in a wrong way to comply to the new behavior, instead of fixing it.

Crap. :(  Thanks for catching and fixing that.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_11"></a>|@tomprince set owner to @therve|
|-|-|

  1. There are a number of missing docstrings. `TestAsynchronousFail`. (It would be nice to add them to the other methods of `TracebackHandling` that got touched as well.
  1. The docstring for `test_exception` isn't clear. The test it is running is synchronous (just raises an exception). I think you mean a test using `twisted.trial.unittest.TestCase` which *can* handle asynchronous tests.
  1. The comment in `_trimFrames` needs to be updated and turned into a proper docstring.
  1. What is going to happen in `_trimFrames` is the traceback doesn't look as expected? We should detect this and do something, as well as test this case. (Doing this may have caught the regression.
Please address the above issues and re-submit.


Note for archaeologists:
  * `SynchronousTestCase` uses `twisted.python.util.runWithWarningsSuppressed` (which doesn't support deferreds), and `TestCase` uses `twisted.internet.utils.runWithWarningsSuppresed` which does).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve removed owner|
|-|-|

I've fixed 1-3. I don't think we can do anything for 4, as there are too many different cases that can happen. We already had a test which ought to have caught the regression.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_13"></a>|@tomprince set owner to @therve|
|-|-|

A couple of last things:
  1. As discussed on IRC, we should document that `_trimFrames` gets called with tracebacks other than the ones we expect (primarly when using testcase not from trial).
  1. `_trimFrames` should use `C{}` and `L{}` for code and references, and have `@param` and `@return`
  1. `getErrorFrames` should have `@param` and `@return`.
Please address the above issues and merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve set status to `closed`|
|-|-|

(In [37172]) Merge async-trim-frames-6282

Author: therve
Reviewer: tom.prince
Fixes: [#6282](https://github.com/twisted/twisted/issues/6282)

Fix a regression in Reporter._trimFrames where utils.runWithWarningsSuppressed
unexpectedly appeared.
